sql/src/main/java/io/crate/planner/operators/WindowAgg.java

```diff
@@ -0,0 +1,151 @@
+                                                      "Only empty OVER() windows are supported. ");
+            }
+        }
+
+        return (tableStats, usedBeforeNextFetch) -> {
+            HashSet<Symbol> allUsedColumns = new HashSet<>(usedBeforeNextFetch);
+            Set<Symbol> columnsUsedInFunctions = extractColumns(windowFunctions);
+            allUsedColumns.addAll(columnsUsedInFunctions);
+            LogicalPlan sourcePlan = source.build(tableStats, allUsedColumns);
+
+            return new WindowAgg(sourcePlan, windowFunctions);
+        };
+    }
+
+    private WindowAgg(LogicalPlan source, List<WindowFunction> windowFunctions) {
+        super(source);
+        this.windowFunctions = windowFunctions;
+    }
+
+    public List<WindowFunction> windowFunctions() {
+        return windowFunctions;
+    }
+
+    @Override
+    protected LogicalPlan updateSource(LogicalPlan newSource, SymbolMapper mapper) {
+        return new WindowAgg(newSource, windowFunctions);
+    }
+
+    @Override
+    public ExecutionPlan build(PlannerContext plannerContext,
+                               ProjectionBuilder projectionBuilder,
+                               int limit,
+                               int offset,
+                               @Nullable OrderBy order,
+                               @Nullable Integer pageSizeHint,
+                               Row params,
+                               SubQueryResults subQueryResults) {
+        ExecutionPlan sourcePlan = source.build(
+            plannerContext,
+            projectionBuilder,
+            limit,
+            offset,
+            order,
+            pageSizeHint,
+            params,
+            subQueryResults
+        );
+
+        sourcePlan.addProjection(new WindowAggProjection(groupFunctionsByWindow(source.outputs(), windowFunctions)));
+        return Merge.ensureOnHandler(sourcePlan, plannerContext);
+    }
+
+    private Map<WindowDefinition, List<Symbol>> groupFunctionsByWindow(List<Symbol> inputs, List<WindowFunction> windowFunctions) {
+        InputColumns.SourceSymbols sourceSymbols = new InputColumns.SourceSymbols(inputs);
+        Map<WindowDefinition, List<Symbol>> groupedFunctions = new HashMap<>();
+
+        for (WindowFunction windowFunction : windowFunctions) {
+            WindowDefinition windowDefinition = windowFunction.windowDefinition();
+
+            Symbol windowFunctionSymbol = InputColumns.create(windowFunction, sourceSymbols);
+            List<Symbol> functions = groupedFunctions.get(windowDefinition);
+            if (functions == null) {
+                ArrayList<Symbol> functionsForWindow = new ArrayList<>();
+                functionsForWindow.add(windowFunctionSymbol);
+                groupedFunctions.put(windowDefinition, functionsForWindow);
+            } else {
+                functions.add(windowFunctionSymbol);
+            }
+        }
+        return ImmutableMap.copyOf(groupedFunctions);
```

Comment:
`Collections.unmodifiableMap` would be cheaper (it's a view instead of a full copy)